code/modules/reagents/machinery/grinder.dm

```diff
@@ -152,108 +152,67 @@
-	for(var/option in options)
-		options[option] = image(icon = 'icons/obj/chemical.dmi', icon_state = "[option]")
-	var/selected
-	selected = show_radial_menu(user, src, options, radius = 42)
-	if(!selected)
-		return
-	switch(selected)
-		if ("grind")
-			grind()
-		if("eject")
-			eject()
-		if ("detach")
-			detach()
-		if("menu")
-			interact(user)
-
-/obj/machinery/reagentgrinder/interact(mob/user as mob) // The microwave Menu
-	if(inoperable())
+/obj/machinery/reagentgrinder/attack_hand(mob/user)
+	. = ..()
+	if(.)
 		return
+
 	user.set_machine(src)
-	var/is_chamber_empty = 0
-	var/is_beaker_ready = 0
-	var/processing_chamber = ""
-	var/beaker_contents = ""
-	var/dat = ""
-
-	if(!inuse)
-		for (var/obj/item/O in holdingitems)
-			processing_chamber += "\A [O.name]<BR>"
-
-		if (!processing_chamber)
-			is_chamber_empty = 1
-			processing_chamber = "Nothing."
-		if (!beaker)
-			beaker_contents = "<B>No beaker attached.</B><br>"
-		else
-			is_beaker_ready = 1
-			beaker_contents = "<B>The beaker contains:</B><br>"
-			var/anything = 0
-			for(var/datum/reagent/R in beaker.reagents.reagent_list)
-				anything = 1
-				beaker_contents += "[R.volume] - [R.name]<br>"
-			if(!anything)
-				beaker_contents += "Nothing<br>"
-
-
-		dat = {"
-	<b>Processing chamber contains:</b><br>
-	[processing_chamber]<br>
-	[beaker_contents]<hr>
-	"}
-		if (is_beaker_ready && !is_chamber_empty && !(stat & (NOPOWER|BROKEN)))
-			dat += "<A href='?src=\ref[src];action=grind'>Process the reagents</a><BR>"
-		if(holdingitems && holdingitems.len > 0)
-			dat += "<A href='?src=\ref[src];action=eject'>Eject the reagents</a><BR>"
-		if (beaker)
-			dat += "<A href='?src=\ref[src];action=detach'>Detach the beaker</a><BR>"
-	else
-		dat += "Please wait..."
-	user << browse("<HEAD><TITLE>All-In-One Grinder</TITLE></HEAD><TT>[dat]</TT>", "window=reagentgrinder")
-	onclose(user, "reagentgrinder")
-	return
+	ui_interact(user)
+
+/obj/machinery/reagentgrinder/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = NANOUI_FOCUS)
+	var/list/data = ui_data()
 
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "grinder.tmpl", name, 400, 550)
+		ui.set_initial_data(data)
+		ui.open()
+
+/obj/machinery/reagentgrinder/ui_data()
+	var/list/data = list()
+	data["on"] = inuse
+
+	if(beaker)
+		data["beaker"] = beaker.reagents.ui_data()
+	data["contents"] = list()
+	for (var/obj/item/O in holdingitems)
+		data["contents"] += "\A [O.name]"
+	return data
 
 /obj/machinery/reagentgrinder/Topic(href, href_list)
 	if(..())
 		return 1
 
-	switch(href_list["action"])
-		if ("grind")
-			grind()
-		if("eject")
-			eject()
-		if ("detach")
-			detach()
+	if(href_list["grind"])
+		grind(usr)
+	if(href_list["eject"])
+		eject(usr)
+	if(href_list["detach"])
+		detach()
 	playsound(loc, 'sound/machines/machine_switch.ogg', 100, 1)
-	src.updateUsrDialog()
+	SSnano.update_uis(src)
 	return 1
```

Comment:
I don't think `update_uis` is necessary. If you return 1, the code above updates it for you.